Improve reliability of memory stats

RandomShaper opened the pull request:
This PR is two-fold:

### 1) Improve reliability of mem stats

- The allocation count is managed atomically and where it actually should change (for instance, not counting an allocation before its success has been checked).

### 2) Memory block byte-tagging

- Memory just allocated or the portion of memory added to a block with a realloc is initialized with some magic number. _Helps catching bugs about unitialized members._
- Memory about to be freed or the portion of a memory block which is about to disappear by a shrinking realloc is filled with a differnt magic number. _Helps catching bugs due to dangling pointers._

Disabled by default. Can be enabled for debug builds by defining `DEBUG_MEMORY_TAGGING`.

reduz commented:
Some feedback about this

1) We should not use mutexes for this, instead I think atomic counters make more sense and are more efficient. I'm not sure if we have 64 bit functions for this, but we could add them in safe_refcount.h like
atomic_increment64
atomic_decrement64

2) Clearing the mem using a pattern, i'm not really sure how much this affects performance..

RandomShaper commented:
1. Lock-free programming is more efficient indeed. I did it that way because that's how it was being done in 2.1. But given that this affects the editor, I agree we should avoid mutex locking.
What do you think about doing this?:
- changing the code to Godot-style atomics;
- adding a mutex-based approach as a fallback for arch we have no true atomic implementation for;
- opening an issue about adding those missing implementations so we don't forget.
(Ah, if we only could use C++11…).

2. In this case, although it affects the editor as well, given it doesn't affect games in release-mode, it's a good thing as it helps catching bugs due to uninitialized data, dangling pointers, etc. In fact, I would even extend it to allocs and reallocs.

bojidar-bg commented on core/os/memory.cpp:
```diff
@@ -142,21 +171,25 @@ void Memory::free_static(void *p_ptr, bool p_pad_align) {
 	} else {
-
+#if !defined DEBUG_ENABLED
+#error "Debug builds expect prepadding to be enabled."
```
I feel this is going to error out on all non-debug builds, as it is a preprocessor directive.

RandomShaper replied:
Definitely. I will fix it for the next push.

RandomShaper committed:
Add atomics for 64-bit integers

RandomShaper committed:
Implement byte-tagging of memory

Using a magic value for memory allocated but uninitialized and another one for memory about-to-be-released. Helps in debugging unitialized members, dangling pointerts, etc.

Disabled by default. Can be enabled for debug builds by defining `DEBUG_MEMORY_TAGGING`.

RandomShaper commented:
Hey, people. I have reworked this PR. Please review it now. Thanks!

RandomShaper commented:
Sorry for spamming but my edit didn't get saved. Please read the PR description now. I've tried to explain things there.

reduz commented:
I still don't understand why the mutex is used, as far as I can see it could all use atomics. Keep in mind most people develop in debug, so forcing a mutex in there even for debug is expensive.

reduz commented:
regarding byte tagging, I would make it optional as it could be pretty expensive with large/medium memory allocs. Just add some static boolean and make it turn on/off using GLOBAL_DEF() in main.cpp

RandomShaper commented:
Well, I was guessing alloc count and memory usage make together a piece of information you can query to check the engine status. In that case we would need the critical section to avoid race conditions.

But I've missed the fact that they are being queried independently so that's probably why you think there is no point in locking. Am I right?

Now, about byte tagging, I intended it as a kind of dev tool to catch bugs. Making it an editor setting looks interesting, but it would be useful only if standard users enabled it so they could be bitten by more bugs and could report them. But how would we encourage them to use it if it has a performance cost?

reduz commented:
They are together a piece of information, but in this case priority is
performance. given cache is synced in the atomic locks and that both values
pretty much likely occupy a single cache line, I think it's  unlikely you
will get a separate reading from them.

Standard users don't really have to gain much from the memory tagging, that
seems like a developer-only tool and honestly we didn't have many
situations where this would help. Also windows already does tag memory when
freed in debug afaik. As this still hits performance, I think it loses on
the cost/benefit vs not having it or having it activated by default.


On Thu, Jun 22, 2017 at 7:16 PM, Pedro J. Estébanez <
notifications@github.com> wrote:

> Well, I was guessing alloc count and memory usage make together a piece of
> information you can query to check the engine status. In that case we would
> need the critical section to avoid race conditions.
>
> But I've missed the fact that they are being queried independently so
> that's probably why you think there is no point in locking. Am I right?
>
> Now, about byte tagging, I intended it as a kind of dev tool to catch
> bugs. Making it an editor setting looks interesting, but it would be useful
> only if standard users enabled it so they could be bitten by more bugs and
> could report them. But how would we encourage them to use it if it has a
> performance cost?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/9014#issuecomment-310517872>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF-Z247By3IkWBfgqldJ30Ga5Gjxh1-eks5sGufTgaJpZM4Nts_0>
> .
>

RandomShaper commented:
The performance argument wins so no need to keep discussing.

About byte tagging, making it an user-selectable setting was your idea, not mine. :)

Let's keep it only for developers, for which I think it's really useful. In my physics fixing PR I catched some bugs thanks to that and I'd have catched another one more easily if I'd had that earlier. So, if you agree, I'll keep it disabled by default but enableable by a define for custom builds.

I will opt-in myself! 8)

reduz commented:
if it's useful for you feel free to implement it, i can imagine it makes
sense in some platforms

On Thu, Jun 22, 2017 at 7:56 PM, Pedro J. Estébanez <
notifications@github.com> wrote:

> The performance argument wins so no need to keep discussing.
>
> About byte tagging, making it an user-selectable setting was your idea,
> not mine. :)
>
> Let's keep it only for developers, for which I think it's really useful.
> In my physics fixing PR I catched some bugs thanks to that and I'd have
> catched another one more easily if I'd had that earlier. So, if you agree,
> I'll keep it disabled by default but enableable by a define for custom
> builds.
>
> I will opt-in myself! 8)
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/9014#issuecomment-310524669>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF-Z25AaVeJ9AKWaCd6-HXdxMSkIXz68ks5sGvETgaJpZM4Nts_0>
> .
>

RandomShaper committed:
Improve reliability of memory stats

The allocation count is managed atomically and where it actually should
change (for instance, not counting an allocation before its success has
been checked).

Bonus: Improve readability of the pre-pad checks.

RandomShaper commented:
I've edited this based on the conversation.

Byte tagging is now disabled by default but it can be enabled with a define. As I think this will help catching bugs, @akien-mga, what do you think about putting a notice somewhere (blog, etc.) encouraging contributors to have it enabled?

akien-mga commented on core/os/memory.cpp:
```diff
@@ -43,41 +44,36 @@ void *operator new(size_t p_size, void *(*p_allocfunc)(size_t p_size)) {
 size_t Memory::max_usage = 0;
+#define PREPAD true
+#else
+#define PREPAD p_pad_align
```
That's bogus no? `p_pad_align` is a parameter of `void *Memory::alloc_static(size_t p_bytes, bool p_pad_align)`, I doubt you can use it in the outer scope for a compiler define.

akien-mga replied:
Or does it get replaced "as is" and it gets evaluated properly as it's by chance only used in the scope where `p_pad_align` exists? In this case, I have the feeling that the previous code with a local boolean was clearer and less error-prone, but maybe this change improves performance?

(BTW those are genuine questions from a less-than-average C++ dev, so feel free to tell me if I'm wrong because I don't understand an underlying concept :D).

RandomShaper replied:
Your last comments gets it right. The preprocessor just replace the defines so the compiler will actually see one thing or the other acording to the build type.

Now, to defend my commit :), the previous code forced prepad in debug builds and anyway you had to review it carefully to understand it.

I wouldn't write such a "robotic" code in other areas, but memory management is low level stuff and anyone wanting to touch this would have anyway to take some time to wrap his head around it (as I had to do).

As an advantage, the reviewed code block shows pretty clearly that things work different across build types and gives some clues about how to interpret the functions below.

akien-mga replied:
Thanks for the explanations, it looks good to me then :)

akien-mga commented:
Please don't merge your own PRs @RandomShaper, unless they're trivial enough to be consensual.

reduz commented:
Sorry, reverted. I thought we agreed this PR was simply adding atomic refcounting to the memory stats to avoid multi threading related problems. It should only focus on that, and the merged PR does not even solve the problem.

reduz commented on core/os/memory.cpp:
```diff
@@ -43,41 +44,36 @@ void *operator new(size_t p_size, void *(*p_allocfunc)(size_t p_size)) {
-	void *mem = malloc(p_bytes + (prepad ? PAD_ALIGN : 0));
-
-	alloc_count++;
+	void *mem = malloc(p_bytes + (PREPAD ? PAD_ALIGN : 0));
```
This code was fine as-is, please don't change out of the blue

reduz commented on core/os/memory.cpp:
```diff
@@ -43,41 +44,36 @@ void *operator new(size_t p_size, void *(*p_allocfunc)(size_t p_size)) {
 
+	if (PREPAD) {
 #ifdef DEBUG_ENABLED
 		mem_usage += p_bytes;
```
this is where atomic ref-couting should be used to avoid multithreading issues. If two CPU cores allocate at the same time, the mem_usage variable might be corrupted

RandomShaper replied:
Then I think I'll try to implement atomic addition and use it for this.

reduz commented on core/os/memory.cpp:
```diff
@@ -138,27 +126,16 @@ void Memory::free_static(void *p_ptr, bool p_pad_align) {
-
-		free(mem);
 	}
+	atomic_decrement(&alloc_count);
```
this one is alright, but it should happen to mem_usage too. You could change it to uint64_t without problems I guess

RandomShaper commented:
Ah, sorry about that. I thought we reached consensus.

I implemented the allocation count as an atomic. The total allocated size was kept as it is since it would probably be atomic on most platforms. I thought that was the best way after removing the critical section.

The byte tagging thing was made optional and kept as a separate commit to have it isolated.

What shall I do to then?

RandomShaper commented:
Never mind. I've just noticed your comments.

reduz commented:
What i don't understand if there is any need to change the prepad code if other than being cosmetic. If so, please leave as is as cosmetic code changes should go in a separate PR, as they are a separate issue to discuss.

RandomShaper commented:
While I was working on this, that made sense to me. It will now be quite difficult to unmix. If you can give me your opinion now about the (in)convienience of it, it'd be very helpful.

Next time I'll split contributions as involved as this into smaller pieces.

reduz commented:
I understand, but I honestly prefer the code the way it is, specially because it's a pretty critical piece of code. If you could do the effort of just submitting a PR that only changes the operators to atomic functions that would be great. If it's too much work for you, please let me know and I will do myself

RandomShaper commented:
No problem. I'll do it!

reduz commented:
Also sorry, did not intend to sound harsh, core/ is the area I take the most care with and any try to scrutinize every change as much as possible before merging, as mistakes there can cost a lot.

RandomShaper commented:
No offense taken! A revert is a painful thing, but I understand your reasons.

Thank you anyway for your words!
